twine/utils.py

```diff
@@ -99,19 +101,44 @@ def get_config(path: str = "~/.pypirc") -> Dict[str, RepositoryConfig]:
     return dict(config)
 
 
+def validate_url(repository_url: Optional[str]) -> bool:
+    """Validate the given url for allowed schemes and components"""
+
+    if not repository_url:
+        return False
+
+    # Scheme should always be https, and the url should at minimum
+    # contain scheme and host
+    validator = (
+        validators.Validator()
+        .allow_schemes("https")
+        .require_presence_of("scheme", "host")
+    )
+
+    # Only return scheme when the url is valid
+    url = uri_reference(repository_url)
+    try:
+        validator.validate(url)
+    except Exception:
+        return False
+
+    return True
+
+
 def get_repository_from_config(
     config_file: str, repository: str, repository_url: Optional[str] = None
 ) -> RepositoryConfig:
     # Get our config from, if provided, command-line values for the
     # repository name and URL, or the .pypirc file
-    if repository_url and "://" in repository_url:
+    scheme = validate_url(repository_url)
```

Comment:
This is no longer representing the scheme but the validity of the whole URL (which includes, whether the URL is truthy or not).

We can simplify everything below if we handled the exception from `validators.validate` ourselves in a far better way by handling the specific exceptions detailed in https://rfc3986.readthedocs.io/en/latest/api-ref/validators.html#rfc3986.validators.Validator.validate.

As it stands, `repository_url and scheme` is borderline in it's usefulness because `validate_url` returns `False` if `not repository_url` and only returns true if `repository_url and validators.validate(url)` (in pseudo-code)

The condition below this could be improved by handling specific exceptions because we can also then detect a missing host and raise an exception about that too.

Reply:
Thanks for the tip. I found that only `MissingComponentError` and `UnpermittedComponentError` are applicable in our scenario. Given that information, I have two ways of handling this.

1. Catch and raise exceptions in `validate_url` and just raise them. Then catch them in `get_repository_from_config` again, and create necessary messages to add to `exceptions.UnreachableRepositoryURLDetected` (I would also change `"Repository URL has missing components. Please add 'https://'. \n".format(repository_url)` to be more generic` to better report about protocol errors.

2. Don't raise any errors in `validate_url`. Instead catch and raise exceptions in `get_repository_from_config` and create necessary messages.

There is a `try/except` block after this which calls `return get_config(config_file)[repository]` . After I add the specific error messages and exceptions in this section, what will be the cases when we would return the dictionary 
`{ "repository": repository_url, "username": None, "password": None}` and when we would be calling the next `try/catch` block under the updated url validation ?

Reply:
I ended up taking the second approach of catching and raising exceptions in `get_repository_from_config` with necessary messages for each exception, and updated `get_repository_from_config` to only call `validate_url` if `repository_url` is not None.